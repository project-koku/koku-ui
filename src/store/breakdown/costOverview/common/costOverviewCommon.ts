import { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum CostOverviewWidgetType {
  cluster = 'cluster', // This type displays clusters associated with a project
  cost = 'cost', // This type displays a cost breakdown as a pie chart
  cpuUsage = 'cpuUsage', // This type displays cpu usage as a bullet chart
  memoryUsage = 'memoryUsage', // This type displays memory usage as a bullet chart
  reportSummary = 'summary', // This type displays a cost report summary
  volumeUsage = 'volumeUsage', // This type displays volume usage as a bullet chart
}

export interface CostOverviewWidget {
  chartName?: string; // Will be the prefix for ids within the chart
  id: number;
  cluster?: {
    reportGroupBy: string; // Report group_by
    showWidgetOnGroupBy?: string[]; // Show widget when group_by is matched
  };
  usage?: {
    showCapacityOnGroupBy?: string[]; // Show capacity when group_by is matched
  };
  reportSummary?: {
    reportGroupBy: string; // Report group_by
    showWidgetOnGroupBy?: string[]; // Show widget when group_by is matched
    usePlaceholder?: boolean; // Use placeholder to keep card placement when widget is not shown
  };
  reportPathsType: ReportPathsType; // Report URL path
  reportType: ReportType; // Report type; cost, storage, etc.
  type: CostOverviewWidgetType;
}
