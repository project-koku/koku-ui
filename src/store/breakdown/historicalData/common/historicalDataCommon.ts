import { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum HistoricalDataWidgetType {
  cost = 'cost', // This type displays historical cost chart
  trend = 'trend', // This type displays historical trend chart
  usage = 'usage', // This type displays historical usage chart
}

export interface HistoricalDataWidget {
  chartName: string; // Will be the prefix for ids within the chart
  id: number;
  reportPathsType: ReportPathsType; // Report URL path
  reportType: ReportType; // Report type; cost, storage, etc.
  type: HistoricalDataWidgetType;
}
