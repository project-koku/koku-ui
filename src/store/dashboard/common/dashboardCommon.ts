import { MessageDescriptor } from '@formatjs/intl/src/types';
import { ForecastPathsType, ForecastType } from 'api/forecasts/forecast';
import { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum DashboardChartType {
  cost = 'cost', // // This displays cumulative cost compared to infrastructure cost
  dailyCost = 'dailyCost', // This displays cumulative and daily cost compared to infrastructure cost
  dailyTrend = 'dailyTrend', // This displays cumulative and daily cost
  trend = 'trend', // This displays cumulative cost only
  usage = 'usage', // This displays daily usage and requests
}

export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget<T> {
  availableTabs?: T[];
  chartType?: DashboardChartType;
  currentTab?: T;
  details: {
    adjustContainerHeight?: boolean; // Adjust chart container height for responsiveness
    costKey?: MessageDescriptor; // i18n key
    formatOptions: ValueFormatOptions;
    requestFormatOptions?: {
      fractionDigits?: number;
    };
    requestKey?: MessageDescriptor;
    showHorizontal?: boolean; // Show horizontal layout
    showTooltip?: boolean; // Show cost tooltip
    showUnits?: boolean; // Show units
    showUsageFirst?: boolean; // Show usage before cost
    showUsageLegendLabel?: boolean;
    units?: string; // Override units shown as workaround for missing Azure API units
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: MessageDescriptor; // i18n key
    viewAllPath?: string; // View all link to details page
  };
  filter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  forecastPathsType?: ForecastPathsType;
  forecastType?: ForecastType;
  id: number;
  reportPathsType: ReportPathsType;
  reportType: ReportType;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: MessageDescriptor;
  tabsFilter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  trend: {
    computedForecastItem?: string; // The computed forecast item to use in charts.
    computedForecastInfrastructureItem?: string; // The computed forecast infrastructure item to use in charts.
    computedReportItem: string; // The computed report item to use in charts, summary, etc.
    computedReportItemValue: string; // The computed report value (e.g., raw, markup, total, or usage)
    dailyTitleKey?: MessageDescriptor;
    showInfrastructureLabel?: boolean; // Trend chart legend items show "Infrastructure cost" instead of "cost"
    showSupplementaryLabel?: boolean; // Trend chart legend items show "Supplementary cost" instead of "cost"
    titleKey: MessageDescriptor;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems?: {
    formatOptions: any;
  };
}
