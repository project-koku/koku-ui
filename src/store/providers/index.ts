import {
  awsProvidersQuery,
  azureProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ociProvidersQuery,
  ocpProvidersQuery,
  providersQuery,
} from 'store/providers/providersCommon';

import * as providersActions from './providersActions';
import { stateKey as providersStateKey } from './providersCommon';
import { ProvidersAction, providersReducer, ProvidersState } from './providersReducer';
import * as providersSelectors from './providersSelectors';

export {
  awsProvidersQuery,
  azureProvidersQuery,
  ociProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ocpProvidersQuery,
  ProvidersAction,
  providersActions,
  providersQuery,
  providersReducer,
  providersSelectors,
  ProvidersState,
  providersStateKey,
};
