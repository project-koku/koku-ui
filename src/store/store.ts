import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';

import { rootReducer, RootState } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
