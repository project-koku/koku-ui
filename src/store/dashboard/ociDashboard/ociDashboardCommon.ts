import { getQuery, OciFilters, OciQuery } from 'api/queries/ociQuery';
import { DashboardWidget } from 'store/dashboard/common/dashboardCommon';

export const ociDashboardStateKey = 'ociDashboard';
export const ociDashboardDefaultFilters: OciFilters = {
  time_scope_units: 'month',
  time_scope_value: -1,
  resolution: 'daily',
};
export const ociDashboardTabFilters: OciFilters = {
  ...ociDashboardDefaultFilters,
  limit: 3,
};

// eslint-disable-next-line no-shadow
export const enum OciDashboardTab {
  product_services = 'product_services',
  payer_tenant_ids = 'payer_tenant_ids',
  regions = 'regions',
}

export interface OciDashboardWidget extends DashboardWidget<OciDashboardTab> {}

export function getGroupByForTab(widget: OciDashboardWidget): OciQuery['group_by'] {
  switch (widget.currentTab) {
    case OciDashboardTab.product_services:
      return {
        product_service:
          widget.tabsFilter && widget.tabsFilter.product_service ? widget.tabsFilter.product_service : '*',
      };
    case OciDashboardTab.payer_tenant_ids:
      return { payer_tenant_id: '*' };
    case OciDashboardTab.regions:
      return { region: '*' };
    default:
      return {};
  }
}

export function getQueryForWidget(filter: OciFilters = ociDashboardDefaultFilters, props?) {
  const query: OciQuery = {
    filter,
    ...(props ? props : {}),
  };
  return getQuery(query);
}

export function getQueryForWidgetTabs(widget: OciDashboardWidget, filter: OciFilters = ociDashboardDefaultFilters) {
  const group_by = getGroupByForTab(widget);
  const newFilter = {
    ...JSON.parse(JSON.stringify(filter)),
  };

  if (
    widget.currentTab === OciDashboardTab.product_services &&
    widget.tabsFilter &&
    widget.tabsFilter.product_service
  ) {
    newFilter.service = undefined;
  }
  const query: OciQuery = {
    filter: newFilter,
    group_by,
  };
  return getQuery(query);
}
