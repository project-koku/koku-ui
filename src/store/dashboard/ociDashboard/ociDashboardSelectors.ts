import { RootState } from 'store/rootReducer';

import {
  getQueryForWidget,
  getQueryForWidgetTabs,
  ociDashboardDefaultFilters,
  ociDashboardStateKey,
  ociDashboardTabFilters,
} from './ociDashboardCommon';

export const selectOciDashboardState = (state: RootState) => state[ociDashboardStateKey];

export const selectWidgets = (state: RootState) => selectOciDashboardState(state).widgets;

export const selectWidget = (state: RootState, id: number) => selectWidgets(state)[id];

export const selectCurrentWidgets = (state: RootState) => selectOciDashboardState(state).currentWidgets;

export const selectWidgetQueries = (state: RootState, id: number) => {
  const widget = selectWidget(state, id);

  const filter = {
    ...ociDashboardDefaultFilters,
    ...(widget.filter ? widget.filter : {}),
  };
  const tabsFilter = {
    ...ociDashboardTabFilters,
    ...(widget.tabsFilter ? widget.tabsFilter : {}),
  };

  return {
    previous: getQueryForWidget({
      ...filter,
      time_scope_value: -2,
    }),
    current: getQueryForWidget(filter),
    forecast: getQueryForWidget({}, { limit: 31 }),
    tabs: getQueryForWidgetTabs(widget, {
      ...tabsFilter,
      resolution: 'monthly',
    }),
  };
};
