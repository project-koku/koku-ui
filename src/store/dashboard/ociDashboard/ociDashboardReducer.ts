import { ActionType, getType } from 'typesafe-actions';

import { setWidgetTab } from './ociDashboardActions';
import { OciDashboardWidget } from './ociDashboardCommon';
import {
  costSummaryWidget,
  databaseWidget,
  networkWidget,
  storageWidget,
  virtualMachineWidget,
} from './ociDashboardWidgets';

export type OciDashboardAction = ActionType<typeof setWidgetTab>;

export type OciDashboardState = Readonly<{
  widgets: Record<number, OciDashboardWidget>;
  currentWidgets: number[];
}>;

export const defaultState: OciDashboardState = {
  currentWidgets: [
    costSummaryWidget.id,
    virtualMachineWidget.id,
    storageWidget.id,
    networkWidget.id,
    databaseWidget.id,
  ],
  widgets: {
    [costSummaryWidget.id]: costSummaryWidget,
    [virtualMachineWidget.id]: virtualMachineWidget,
    [databaseWidget.id]: databaseWidget,
    [networkWidget.id]: networkWidget,
    [storageWidget.id]: storageWidget,
  },
};

export function ociDashboardReducer(state = defaultState, action: OciDashboardAction): OciDashboardState {
  switch (action.type) {
    case getType(setWidgetTab):
      return {
        ...state,
        widgets: {
          ...state.widgets,
          [action.payload.id]: {
            ...state.widgets[action.payload.id],
            currentTab: action.payload.tab,
          },
        },
      };
    default:
      return state;
  }
}
