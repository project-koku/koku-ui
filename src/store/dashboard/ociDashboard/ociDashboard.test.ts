jest.mock('store/reports/reportActions');

import { ReportType } from 'api/reports/report';
import { ChartType } from 'pages/views/components/charts/common/chartDatumUtils';
import { createMockStoreCreator } from 'store/mockStore';
import { reportActions } from 'store/reports';

import * as actions from './ociDashboardActions';
import { getGroupByForTab, getQueryForWidgetTabs, ociDashboardStateKey, OciDashboardTab } from './ociDashboardCommon';
import { ociDashboardReducer } from './ociDashboardReducer';
import * as selectors from './ociDashboardSelectors';
import {
  costSummaryWidget,
  databaseWidget,
  networkWidget,
  storageWidget,
  virtualMachineWidget,
} from './ociDashboardWidgets';

const createOciDashboardStore = createMockStoreCreator({
  [ociDashboardStateKey]: ociDashboardReducer,
});

const fetchReportMock = reportActions.fetchReport as jest.Mock;

beforeEach(() => {
  fetchReportMock.mockReturnValue({ type: '@@test' });
});

test('default state', () => {
  const store = createOciDashboardStore();
  const state = store.getState();
  expect(selectors.selectCurrentWidgets(state)).toEqual([
    costSummaryWidget.id,
    virtualMachineWidget.id,
    storageWidget.id,
    networkWidget.id,
    databaseWidget.id,
  ]);
  expect(selectors.selectWidget(state, costSummaryWidget.id)).toEqual(costSummaryWidget);
});

test('fetch widget reports', () => {
  const store = createOciDashboardStore();
  store.dispatch(actions.fetchWidgetReports(costSummaryWidget.id));
  expect(fetchReportMock.mock.calls).toMatchSnapshot();
});

test('changeWidgetTab', () => {
  const store = createOciDashboardStore();
  store.dispatch(actions.changeWidgetTab(costSummaryWidget.id, OciDashboardTab.resource_locations));
  const widget = selectors.selectWidget(store.getState(), costSummaryWidget.id);
  expect(widget.currentTab).toBe(OciDashboardTab.resource_locations);
  expect(fetchReportMock).toHaveBeenCalledTimes(3);
});

describe('getGroupByForTab', () => {
  test('services tab', () => {
    const widget = getGroupByForTab({
      currentTab: OciDashboardTab.service_names,
    });
    expect(widget).toMatchSnapshot();
  });

  test('accounts tab', () => {
    const widget = getGroupByForTab({
      currentTab: OciDashboardTab.payer_tenant_ids,
    });
    expect(widget).toMatchSnapshot();
  });

  test('regions tab', () => {
    const widget = getGroupByForTab({
      currentTab: OciDashboardTab.resource_locations,
    });
    expect(widget).toMatchSnapshot();
  });

  test('unknown tab', () => {
    expect(getGroupByForTab('unknown' as any)).toMatchSnapshot();
  });
});

test('getQueryForWidget', () => {
  const widget = {
    id: 1,
    titleKey: '',
    reportType: ReportType.cost,
    availableTabs: [OciDashboardTab.payer_tenant_ids],
    currentTab: OciDashboardTab.payer_tenant_ids,
    details: { formatOptions: {} },
    trend: {
      formatOptions: {},
      titleKey: '',
      type: ChartType.daily,
    },
    topItems: {
      formatOptions: {},
    },
  };

  [
    [
      undefined,
      'filter[time_scope_units]=month&filter[time_scope_value]=-1&filter[resolution]=daily&group_by[payer_tenant_id]=*',
    ],
    [{}, 'group_by[payer_tenant_id]=*'],
    [{ limit: 3 }, 'filter[limit]=3&group_by[payer_tenant_id]=*'],
  ].forEach(value => {
    expect(getQueryForWidgetTabs(widget, value[0])).toEqual(value[1]);
  });
});
