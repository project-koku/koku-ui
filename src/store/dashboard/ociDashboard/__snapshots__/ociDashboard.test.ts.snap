// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch widget reports 1`] = `
Array [
  Array [
    "oci",
    "cost",
    "filter[time_scope_units]=month&filter[time_scope_value]=-1&filter[resolution]=daily",
  ],
  Array [
    "oci",
    "cost",
    "filter[time_scope_units]=month&filter[time_scope_value]=-2&filter[resolution]=daily",
  ],
  Array [
    "oci",
    "cost",
    "filter[time_scope_units]=month&filter[time_scope_value]=-1&filter[resolution]=monthly&filter[limit]=3&group_by[product_service]=*",
  ],
]
`;

exports[`getGroupByForTab accounts tab 1`] = `
Object {
  "payer_tenant_id": "*",
}
`;

exports[`getGroupByForTab regions tab 1`] = `
Object {
  "region": "*",
}
`;

exports[`getGroupByForTab services tab 1`] = `
Object {
  "product_service": "*",
}
`;

exports[`getGroupByForTab unknown tab 1`] = `Object {}`;
